QuantumRangers-SE-Team-10/lasertech-backend
# Extend Database with Game and PlayerSession Entities & Introduce Related API Controllers

### Summary
In this change, we enable the operator to create a new game and allow players to join a game.

### Details
We start by adding two additional table to the database. The Game table keeps track of the red and blue team score along with an auto incrementing GameID field. The second table is the PlayerSession table, this table keep track of each player session per game. The fields for the PlayerSession table includes the PlayerID, GameID, Team which determine if they are on red or blue team, Score which is the player score for that session, EquipmentID. The Game table have a one to many relationship to the PlayerSession, because each game can have up to 30 players, 30 player session will be link to the same Game entry, the equi join will be on GameID. The PlayerSession have a one to one relationship with the Game table and the Player table joining both via GameID and PlayerID respectively. 

Next I worked on the controller, I created two separate controller. One named GameController which create the API endpoint need to initialize a new game session, retrieve all game session or retrieve a single game session by id. The second controller, PlayerSession controller allow the operator to add a new player to an existing game session.

### Design Decision
I decided to use multiple table to facilitate the functionality of a game session mostly because I want separation of concern and reducing duplicate data in the database. Enum was use instead of a boolean value in order to make Team more adaptable, so instead of having only red and blue team, we can also have a green team and so on. 


### References
[One-to-One Relationship in entity framework core](https://www.youtube.com/watch?v=ZXJNW3SXEe8&list=PLdHN14J7CHtaNHnAPOk_yEq0GidtbTizl&index=10)
[One-to-Many Relationship in entity framework core](https://www.youtube.com/watch?v=2ODijPU_2b0&t=1s)


### Testing instruction
1. Create a new game session by sending a POST request to `/api/game` the server should respond with the new game object, note the GameID
<img width="1512" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/6e8601f5-0970-4662-8a0f-02d8ec345f2d">

2. Retrieve all game session by sending a GET request to `/api/game` the server should respond with the all of the game object including the one you just created.
<img width="1512" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/74f7a217-075a-450f-857d-4e10aeb0d805">

3. Retrieve a game by id, send a get request to `/api/game/{gameID}` passing in id "2" as the param.
<img width="1512" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/9dd89769-a65a-48be-9d87-08d0cdd24143">

The server should respond with the correct game object that contain an array of 2 PlayerSession.

<img width="1512" alt="image" src="https://github.com/QuantumRangers-SE-Team-10/lasertech-backend/assets/76791231/472483c1-d2b9-46c2-84f0-2a9527e42f54">


### Checks
